src/logic_utils/truth_table.cpp

```diff
@@ -0,0 +1,177 @@
+#include <iostream>
+
+#include <string>
+
+#include <stack>
+
+#include <cctype>
+
+#include <unordered_map>
+#include <map>
+#include <vector>
+
+#include <set>
+
+using namespace std;
+
+
+map < int, char > variables;
+
+
+
+bool isOperator(char c) {
+return c == '+' || c == '*' || c == '~' || c == '!';
+}
+
+int getPrecedence(char op) {
+  if (op == '+')
+    return 1;
+  if (op == '*')
+    return 2;
+  return 0;
+}
+
+double applyOperation(double a, double b, char op) {
+  switch (op) {
+  case '+':
+    return a || b;
+  case '*':
+    return a && b;
+  default:
+    return 0.0; // Invalid operator
+  }
+}
+
+double evaluateExpression(const std::string & expression) {
+  std::stack < double > values;
+  std::stack < char > operators;
+
+  for (size_t i = 0; i < expression.length(); ++i) {
+    char c = expression[i];
+
+    if (isspace(c))
+      continue;
+
+    if (isdigit(c)) {
+      std::string num;
+      while (i < expression.length() && (isdigit(expression[i]) || expression[i] == '.')) {
+        num += expression[i];
+        ++i;
+      }
+      --i;
+      values.push(std::stod(num));
+    } else if (c == '(') {
+      operators.push(c);
+    } else if (c == ')') {
+      while (!operators.empty() && operators.top() != '(') {
+        double b = values.top();
+        values.pop();
+
+        double a = values.top();
+        values.pop();
+
+        char op = operators.top();
+        operators.pop();
+
+        values.push(applyOperation(a, b, op));
+      }
+
+      if (!operators.empty())
+        operators.pop(); // Discard the opening parenthesis
+    } else if (isOperator(c)) {
+      while (!operators.empty() && getPrecedence(c) <= getPrecedence(operators.top())) {
+        double b = values.top();
+        values.pop();
+
+        double a = values.top();
+        values.pop();
+
+        char op = operators.top();
+        operators.pop();
+
+        values.push(applyOperation(a, b, op));
+      }
+
+      operators.push(c);
+    }
+  }
+
+  while (!operators.empty()) {
+    double b = values.top();
+    values.pop();
+
+    double a = values.top();
+    values.pop();
+
+    char op = operators.top();
+    operators.pop();
+
+    values.push(applyOperation(a, b, op));
+  }
+  cout << values.top() << endl;
```

Comment:
No cout or any operation with stdin/stdout since the console will not be used, but rather a webpage.